### PR TITLE
a8n: Ensure Campaign is not processing when closing/deleting

### LINKED ISSUE #8241: Campaign edit/close/delete buttons should be disabled while Campaign is processing
We should disable the Edit/Close/Delete buttons on the Campaign details page while the Campaign is still processing and changesets are being created on the codehost.

We already don't allow updating the campaign while it's processing and in https://github.com/sourcegraph/sourcegraph/pull/8240 I added the checks for Close/Delete on the backend side.

### PR DESCRIPTION
**Pull request by mrnugget**
This does two things:

1. It checks that the Campaign to be closed/deleted is not processing
   (changesets are currently being created on the codehost). If a
   Campaign were deleted/closed while being processed we might leave
   changesets open on the code hosts.

2. It makes sure that the transactions in the two methods are committed
   before starting a goroutine that closes the changesets. That removes
   a race condition in which the goroutine would try to close changesets
   even though the transaction could not be committed (due to concurrent
   updates for examples).

Closes #8241

### CODE REVIEW

**Comment by eseliger** [comment]
I will add the required changes for the close/delete ui

**Comment by mrnugget** [comment]
> I will add the required changes for the close/delete ui

See: https://github.com/sourcegraph/sourcegraph/issues/8241 :)

File: enterprise/internal/a8n/service.go [hunk 1]
```diff
@@ -1007,3 +1054,11 @@ func mergeByRepoID(chs []*a8n.ChangesetJob, cas []*a8n.CampaignJob) (map[int32]*
 
 	return jobs, nil
 }
+
+func campaignIsProcessing(ctx context.Context, store *Store, campaign int64) (bool, error) {
+	status, err := store.GetCampaignStatus(ctx, campaign)
+	if err != nil {
+		return false, err
+	}
+	return !status.Finished() && !status.Canceled, nil
```
Comment (eseliger) [inline enterprise/internal/a8n/service.go]: @mrnugget just to confirm I understood the logic in GetCampaignStatus correct, this is the same as in the gql api:
`campaign.status != CANCELED && campaign.status != PROCESSING`, right?

Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: Yeah, although now that I look at the code, the method here could be changed to just check whether `campaign.status.state == PROCESSING`

Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: Did that.

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx [hunk 2]
```diff
@@ -416,14 +433,36 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
...
+                                                    }
                                                     className="position-absolute campaign-details__details-menu"
                                                 />
                                             </details>
                                         )}
                                         <details className="campaign-details__details">
                                             <summary>
-                                                <span className="btn btn-danger dropdown-toggle">Delete</span>
+                                                <span
+                                                    className={classNames(
```
Comment (eseliger) [inline web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: I will clean this up in #8117 when the summary moves in the `CloseDeleteCampaignPrompt` component

File: enterprise/internal/a8n/service.go [hunk 3]
```diff
@@ -456,23 +461,38 @@ func (s *Service) CloseCampaign(ctx context.Context, id int64, closeChangesets b
 		return nil, err
 	}
 
-	defer tx.Done(&err)
```
Comment (ryanslade) [inline enterprise/internal/a8n/service.go]: Can you not do something like:

```go
defer func() {
  tx.Done(&err)
  if err == nil && closeChangesets {
  go ...
}
}()
```

File: enterprise/internal/a8n/service.go [hunk 4]
```diff
@@ -532,11 +557,29 @@ func (s *Service) DeleteCampaign(ctx context.Context, id int64, closeChangesets
 		tr.Finish()
 	}()
 
+	tx, err := s.store.Transact(ctx)
+	if err != nil {
+		return err
+	}
+
```
Comment (ryanslade) [inline enterprise/internal/a8n/service.go]: Same as previous comment, can we defer closing tx and spawning the goroutine?

File: enterprise/internal/a8n/service.go (hunk 3 above)
Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: I need the changesets that are loaded before

Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: I'll try some other pattern, though.

**Pull request merged**